### PR TITLE
feat: support only running a subset of components of a Spin app

### PR DESCRIPTION
**Pull request by kate-goldenring**
Draft until support for executing only a set of components is merged in the shim: https://github.com/spinkube/containerd-shim-spin/pull/197

Spin changes: https://github.com/fermyon/spin/pull/2826.

### CODE REVIEW

File: apps/salutations/spin.toml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+[[trigger.http]]
+route = "/hi"
+component = "hello"
+
+[component.hello]
+source = "../hello-world/main.wasm"
+allowed_outbound_hosts = []
+[component.hello.build]
+command = "pushd ../hello-world && tinygo build -target=wasi -gc=leaking -no-debug -o main.wasm main.go && popd"
+watch = ["**/*.go", "go.mod"]
```
Comment (calebschoepp) [inline apps/salutations/spin.toml:29]: Sarcastic nit: It bothers me that `goodbye` is defined before `hello` 🤣

File: config/samples/component-filtering.yaml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+apiVersion: core.spinoperator.dev/v1alpha1
+kind: SpinApp
+metadata:
+  name: hello-salutation-spinapp
+spec:
+  image: "ghcr.io/kate-goldenring/spin-operator/examples/spin-salutations:20241022-144454"
```
Comment (calebschoepp) [inline config/samples/component-filtering.yaml:6]: A TODO should be left or issue opened to come back and point at the spinkube sample app image once this merges.

**Review by calebschoepp — DISMISSED** [review]
Looking for a few cleanup things but I think this is on the right track!

File: apps/salutations/spin.toml (hunk 1 above)
Comment (kate-goldenring) [reply apps/salutations/spin.toml:29]: now i am bothered 😩

File: internal/controller/spinapp_controller.go [hunk 3]
```diff
@@ -429,6 +430,12 @@ func constructDeployment(ctx context.Context, app *spinv1alpha1.SpinApp, config
 	}
 
 	env := ConstructEnvForApp(ctx, app, spinapp.DefaultHTTPPort, config.Otel)
+	if app.Spec.Components != nil {
+		env = append(env, corev1.EnvVar{
+			Name:  "SPIN_COMPONENTS_TO_RETAIN",
+			Value: strings.Join(app.Spec.Components[:], ","),
+		})
+	}
```
[the remark this answers is not in the document]
Comment (kate-goldenring) [reply internal/controller/spinapp_controller.go:438]: No. Not without pulling down the OCI artifact. Same goes for the `spin-kube` plugin. The shim does this and lifts errors

File: api/v1alpha1/spinapp_types.go [hunk 4]
```diff
@@ -80,6 +80,10 @@ type SpinAppSpec struct {
 
 	// Resources defines the resource requirements for this app.
 	Resources Resources `json:"resources,omitempty"`
+
+	// Components to be executed in this group.
+	// +kubebuilder:validation:MinItems:=1
+	Components []string `json:"components,omitempty"`
```
[the remark this answers is not in the document]
Comment (kate-goldenring) [reply api/v1alpha1/spinapp_types.go:86]: This is out of date. Thank you for catching the "group" wording and adding the default behavior description.

**Comment by kate-goldenring** [comment]
@calebschoep CI is failing with:
```
redis_test.go:165: Failed to create spinapp: admission webhook "vspinapp.kb.io" denied the request: SpinApp.core.spinoperator.dev "test-spintainer-redis" is invalid: spec.executor: Invalid value: "spintainer": executor does not exist in namespace
```

**Review by calebschoepp — COMMENTED** [review]
Thanks for addressing those comments!

**Review by endocrimes — APPROVED** [review]

**Review by michelleN — APPROVED** [review]

File: README.md [hunk 5]
```diff
@@ -56,7 +56,7 @@ Create a k3d cluster:
 
 ```shell
 k3d cluster create wasm-cluster \
-    --image ghcr.io/spinkube/containerd-shim-spin/k3d:v0.15.1 \
+    --image ghcr.io/spinkube/containerd-shim-spin/k3d:v0.16.0 \
```
[the remark this answers is not in the document]
Comment (kate-goldenring) [reply README.md:59]: maybe we start pushing a `latest` tag?

**Comment by kate-goldenring** [comment]
> Could we add an e2e for this (or an issue) as a follow up task

@michelleN sure. I'll add that as a follow up and use it as an opportunity to update the image reference to one in the spinkube GHCR

**Comment by kate-goldenring** [comment]
@michelleN @endocrimes @calebschoepp feel free to merge whenever appropriate. I don't have perms

**Comment by kate-goldenring** [comment]
Added test here: https://github.com/spinkube/spin-operator/pull/328

**Pull request merged**